src/vm/comdelegate.cpp

```diff
@@ -1626,35 +1626,9 @@ FCIMPL3(PCODE, COMDelegate::AdjustTarget, Object* refThisUNSAFE, Object* targetU
     // close delegates
     MethodTable* pMTTarg = target->GetMethodTable();
     MethodTable* pMTMeth = pMeth->GetMethodTable();
-
-    BOOL isComObject = false;
-
-#ifdef FEATURE_COMINTEROP
-    isComObject = pMTTarg->IsComObjectType();
-#endif // FEATURE_COMINTEROP
     
     MethodDesc *pCorrectedMethod = pMeth;
 
-    if (pMTMeth != pMTTarg)
-    {
-        //They cast to an interface before creating the delegate, so we now need 
-        //to figure out where this actually lives before we continue.
-        //<TODO>@perf:  Grovelling with a signature is really slow.  Speed this up.</TODO>
-        if (pCorrectedMethod->IsInterface())
-        {
-            // No need to resolve the interface based method desc to a class based
-            // one for COM objects because we invoke directly thru the interface MT.
-            if (!isComObject)
-            {
-                // <TODO>it looks like we need to pass an ownerType in here.
-                //  Why can we take a delegate to an interface method anyway?  </TODO>
-                // 
-                pCorrectedMethod = pMTTarg->FindDispatchSlotForInterfaceMD(pCorrectedMethod, TRUE /* throwOnConflict */).GetMethodDesc();
-                _ASSERTE(pCorrectedMethod != NULL);
-            }
-        }
-    }
-
```

Comment:
This change looks correct as far as I can tell; however, I am concerned that it probably isn't complete. In particular, I suspect that the same failure occurs when we go down the slow path of delegate construction, which is triggered by COMDelegate::GetDelegateCtor returning NULL.. In particular, when NeedsWrapperDelegate returns true, I believe we hit the problematic code. (To test this you'd need a test running on arm32 as wrapper delegates are now somewhat rare). I would like to see you run the problematic testcase which inspired this fix (and the rest of the test tree) on a build of coreclr where COMDelegate::GetDelegateCtor has been changed so that it returns NULL unconditionally at line 3088, or build test cases which force use of slow path delegate creation in some other way.

Reply:
Yup, your suspicion is correct. After I added the test for this, the Ubuntu arm Cross Checked Innerloop Build and Test leg started failing the test. I'm adding a new commit that deletes the remainder of the suspected cargo cult and we'll see how that works out.

Reply:
Could you write a test that goes down the problematic path even on amd64?

Reply:
`NeedsWrapperDelegate` can only return TRUE on ARM and the other cases when `GetDelegateCtor` returns null don't seem applicable (static methods, or instance methods on `System.Nullable`).

Reply:
At the very least you can run slow path delegate construction logic by using Delegate.CreateDelegate.

Reply:
> At the very least you can run slow path delegate construction logic by using Delegate.CreateDelegate.

OK, added the test, although I don't see how I could make it exercise the problematic path. As far as I can tell, there's no way to create a non-virtual delegate with this API and it will always do the virtual resolution. The bug here was that the normal delegate creation sequence wasn't respecting the difference between ldvirtftn and ldftn and would always do ldvirtftn. `Delegate.CreateDelegate` always does the equivalent of ldvirtftn. I'm pretty sure the newly added test would pass before this fix too.